src/sidebar/helpers/test/thread-sorters-test.js

```diff
@@ -77,100 +77,128 @@ describe('sidebar/util/thread-sorters', () => {
-      },
-      {
-        a: { annotation: { location: cfiLocation('/2/4') } },
-        b: { annotation: { location: cfiLocation('/2/2') } },
-        expected: 1,
-      },
+    it('sorts by page index', () => {
+      const a = thread({ pageIndex: 1 });
+      const b = thread({ pageIndex: 2 });
+      const c = thread({ pageIndex: 2 }, newDate);
 
-      // CFI and position
-      {
-        a: { annotation: { location: cfiLocation('/2/2', 100) } },
-        b: { annotation: { location: cfiLocation('/2/4', 10) } },
-        expected: -1,
-      },
-      {
-        a: { annotation: { location: cfiLocation('/2/4', 100) } },
-        b: { annotation: { location: cfiLocation('/2/4', 10) } },
-        expected: 1,
-      },
-    ].forEach((testCase, index) => {
-      it(`sorts by CFI when present (${index})`, () => {
-        assert.equal(
-          // Disable eslint: `sorters` properties start with capital letters
-          // to match their displayed sort option values
-          /* eslint-disable-next-line new-cap */
-          sorters.Location(testCase.a, testCase.b),
-          testCase.expected,
-        );
+      assert.equal(compareLocation(a, b), -1);
+      assert.equal(compareLocation(a, a), 0);
+      assert.equal(compareLocation(b, a), 1);
+      assert.equal(compareLocation(b, c), -1);
+    });
+
+    it('sorts by distance from top of page', () => {
+      const a = thread({ top: 1 });
+      const b = thread({ top: 2 });
+      const c = thread({ top: 2 }, newDate);
+
+      assert.equal(compareLocation(a, b), -1);
+      assert.equal(compareLocation(a, a), 0);
+      assert.equal(compareLocation(b, a), 1);
+      assert.equal(compareLocation(b, c), -1);
+    });
+
+    it('sorts by character offset', () => {
+      const a = thread(charOffset(5));
+      const b = thread(charOffset(10));
+      const c = thread(charOffset(10), newDate);
+
+      assert.equal(compareLocation(a, b), -1);
+      assert.equal(compareLocation(a, a), 0);
+      assert.equal(compareLocation(b, a), 1);
+      assert.equal(compareLocation(b, c), -1);
+    });
+
+    it('sorts by creation date if annotations do not have comparable locations', () => {
+      const a = thread({}, oldDate);
+      const b = thread({}, newDate);
+
+      assert.equal(compareLocation(a, b), -1);
+      assert.equal(compareLocation(a, a), 0);
+      assert.equal(compareLocation(b, a), 1);
+    });
+
+    it('sorts by CFI > page > position > char offset > creation date', () => {
+      // Create annotations with all location properties, where `a < b` for
+      // each property.
+      const a = thread({
+        cfi: '/2/2',
+        pageIndex: 1,
+        top: 100,
+        charOffset: 10,
       });
+
+      const b = thread(
+        {
+          cfi: '/2/4',
+          pageIndex: 2,
+          top: 101,
+          charOffset: 15,
+        },
+        newDate,
+      );
+
+      const deleteFields = (thread, fields) => {
+        const clone = structuredClone(thread);
```

Comment:
Yay for browsers having a native API for deep cloning objects.

Reply:
One learns something new every day 😅 https://developer.mozilla.org/en-US/docs/Web/API/Window/structuredClone